pkg/querier/block_streaming.go

```diff
@@ -181,41 +184,103 @@ func (s *storeGatewayStreamReader) StartBuffering() {
 
+		// Why do we buffer batches received before we receive the estimated number of chunks?
+		// We want the chunks channel to buffer only one batch, so that during query execution we don't read too many chunks into memory at once.
+		// However, we may receive multiple batches before we receive the estimate, so we can't push all batches into the channel - we don't start
+		// reading series until we've received the estimate, so we might end up deadlocked waiting for the estimate while we can't push the batches
+		// in front of the estimate message into the channel.
+		// Under normal circumstances, there's only one batch before the estimate is received, so the impact of this in terms of peak memory consumption
+		// should be minimal.
+		batchesReceivedBeforeEstimate := make([]*storepb.StreamingChunksBatch, 0, 1)
+		haveReceivedEstimate := false
+
+		sendBatch := func(c *storepb.StreamingChunksBatch) error {
+			select {
+			case <-ctxDone:
+				// Why do we abort if the context is done?
+				// We want to make sure that this goroutine is never leaked.
+				// This goroutine could be leaked if nothing is reading from the buffer, but this method is still trying to send
+				// more series to a full buffer: it would block forever.
+				// So, here, we try to send the series to the buffer if we can, but if the context is cancelled, then we give up.
+				// This only works correctly if the context is cancelled when the query request is complete or cancelled,
+				// which is true at the time of writing.
+				return s.client.Context().Err()
+			case s.seriesChunksChan <- c:
+				// Batch enqueued successfully, nothing else to do for this batch.
+				return nil
+			}
+		}
+
 		for {
 			msg, err := s.client.Recv()
 			if err != nil {
-				if errors.Is(err, io.EOF) {
-					if totalSeries < s.expectedSeriesCount {
-						onError(fmt.Errorf("expected to receive %v series, but got EOF after receiving %v series", s.expectedSeriesCount, totalSeries))
-					} else {
-						level.Debug(log).Log("msg", "finished streaming", "series", totalSeries, "chunks", totalChunks)
-					}
-				} else {
+				if !errors.Is(err, io.EOF) {
 					onError(err)
+					return
+				}
+
+				if totalSeries < s.expectedSeriesCount {
+					onError(fmt.Errorf("expected to receive %v series, but got EOF after receiving %v series", s.expectedSeriesCount, totalSeries))
+					return
 				}
 
+				if !haveReceivedEstimate {
+					level.Warn(log).Log("msg", "did not receive chunks count estimate for stream before EOF")
+					s.chunkCountEstimateChan <- 0
+
+					for _, b := range batchesReceivedBeforeEstimate {
+						if err := sendBatch(b); err != nil {
+							onError(err)
+							return
+						}
+					}
+				}
+
+				level.Debug(log).Log("msg", "finished streaming", "series", totalSeries, "chunks", totalChunks)
 				return
 			}
 
-			c := msg.GetStreamingChunks()
-			if c == nil {
-				onError(fmt.Errorf("expected to receive StreamingSeriesChunks, but got something else"))
```

Comment:
I think the change done in this PR is not backward compatible. Let's say you rollout store-gateways before queriers. The store-gateways start sending the estimated chunks message, then the querier running the old version hit this like and the query fails.

If I'm correct, then I think we should keep the chunks estimation disabled, and when disabled the store-gateway doesn't send the message at all, so that it doesn't break here.

Reply:
Streaming chunks from store-gateways to queriers is an experimental feature that is disabled by default. So my rationale for the backwards-incompatible change is that it'll simplify things in the long run if store-gateways just always send the estimate, and anyone who has streaming chunks enabled can disable it during the upgrade, then enable it once the rollout is complete.

Reply:
Ok fair enough.